### ISSUE TITLE
Working with std::vector<std::string>

### ISSUE THREAD

**Issue by DraTeots** [scikit-hep/uproot#338]
Before I write that things below, thank you very much for your uproot effort!

Now working with `std::vector<std::string>` seems a bit difficult. 

As an example I have a ROOT file where each event has hits information. Something like:

```c++
vector<double> hit_x
vector<double> hit_y
vector<double> hit_z
vector<string> hit_vol_name
```
While all vector<double> work like a charm, the interpretation of `hit_vol_name` is 

```
hit_vol_name        TStreamerSTL               asgenobj(STLVector(STLString()))
hit_x                      TStreamerSTL               asjagged(asdtype('>f8'), 10)
hit_y                      TStreamerSTL               asjagged(asdtype('>f8'), 10)
hit_z                      TStreamerSTL               asjagged(asdtype('>f8'), 10)
```

And it causes problems, when converted to pandas, for example:

![Screenshot from 2019-09-17 10-00-10](https://user-images.githubusercontent.com/4732545/65048473-2db5eb80-d932-11e9-8747-e15b010672cc.png)

And then I have to do the first round of conversions:

```python
from awkward import JaggedArray
import awkward
a = JaggedArray.fromiter(file['events']['hit_vol_name'].array())
```
But then it is byte strings, so like in #228 one has to do something like:

```python
import numpy as np
b = a.flatten().regular().astype('U')
```
In this case flattening was Ok, but... 

It seems both inefficient and inconvenient. Or are there better workarounds?

**Comment by jpivarski** [comment]
Actually, what you did here is correct. There isn't a more efficient way to do it (in Python): the data are stored in a non-columnar way, so Numpy tricks aren't going to work—there must be a for loop at some level, and that's what the `fromiter` is.

I'm currently working on a [C++ version of awkward-array](https://github.com/jpivarski/awkward-1.0#readme) that will at least move the for loop in `fromiter` from Python into C++. Meanwhile, ROOT is experimenting with columnar representations of more-than-singly nested data types as [RNtuple](https://root.cern/doc/master/classROOT_1_1Experimental_1_1RNTuple.html) (a replacement of TTree).

As for convenience, there's a tension between transforming a lot of things under the hood versus being simple so that a user can more easily see what it's doing. The reason for not converting bytestrings into strings was because there might be some pre-unicode character sets in old, non-English ROOT files. (Seems likely, given ROOT's age and international distribution.) However, I've gotten more comments/bug reports about the bytestrings, so I'm thinking of changing the default to UTF-8, a default users with old, international files would have to opt-out of.

As for not automatically running `fromiter` on non-columnar data, that would be a large, up-front performance cost when I don't know if you're going to be accessing all of the elements. Currently, uproot just loads non-columnar data and decodes elements on demand with Python code generated from streamers—calling `fromiter` on it decodes all elements and does the conversion to nested columnar data. If you're dropping everything into Pandas, then clearly you want to do that for every element, but maybe not for other applications, and I can't think of a _simple_ interface that would let someone opt-out of it.

In short, making it too automatic means that it's hard to disable the automatic behavior when you don't want it.

**Comment by jpivarski** [comment]
Actually, you don't need to flatten the array to use Numpy's `astype("U")` (if you need to maintain the jagged structure). You can put a generator comprehension between the array extraction and the `awkward.fromiter` call to turn all the bytestrings into strings without losing the nested structure. Then when `fromiter` picks them up, they'll already be strings.

A third way: I think `fromiter` gives you a `JaggedArray` of `StringArray`, which has an `encoding` property. You might be able to

```python
a.content.encoding = "ascii"    # for your definition of 'a'
```

(It's not a conversion, but a reinterpretation, so be sure you don't have any 8-bit characters!)
